File: apps/tradsphere/api/v1/helpers/broadcastCalendar.py
from __future__ import annotations

from calendar import monthrange
from datetime import date, datetime, timedelta
from typing import Any


def _coerce_to_date(given_date: date | datetime | str) -> date:
    if isinstance(given_date, datetime):
        return given_date.date()
    if isinstance(given_date, date):
        return given_date
    if isinstance(given_date, str):
        text = given_date.strip()
        if not text:
            raise ValueError("givenDate cannot be empty")
        try:
            return date.fromisoformat(text)
        except ValueError as exc:
            raise ValueError("givenDate must be ISO date format YYYY-MM-DD") from exc
    raise TypeError("givenDate must be date, datetime, or ISO date string")


def _monday_of_week(target_date: date) -> date:
    return target_date - timedelta(days=target_date.weekday())


# Broadcast buckets anchor on the week-ending Sunday.
def _calendar_month_start(month: int, year: int) -> date:
    month_start = date(year, month, 1)
    if month_start.weekday() == 0:
        return month_start
    return month_start + timedelta(days=7 - month_start.weekday())


def _calendar_month_end(month: int, year: int) -> date:
    month_end = date(year, month, monthrange(year, month)[1])
    return _monday_of_week(month_end) + timedelta(days=6)


def get_broadcast_month_range(month: int, year: int) -> tuple[date, date]:
    month_start = date(year, month, 1)
    month_end = date(year, month, monthrange(year, month)[1])
    begin_broadcast_month = _monday_of_week(month_start)

    number_of_broadcast_week = ((month_end - begin_broadcast_month).days + 1) // 7
    if number_of_broadcast_week < 1:
        raise ValueError("Unable to compute broadcast month range")

    end_broadcast_month = begin_broadcast_month + timedelta(
        days=number_of_broadcast_week * 7 - 1
    )
    return begin_broadcast_month, end_broadcast_month


def get_calendar_month_range(month: int, year: int) -> tuple[date, date]:
    begin_calendar_month = _calendar_month_start(month, year)
    end_calendar_month = _calendar_month_end(month, year)
    return begin_calendar_month, end_calendar_month


def get_broadcast_month_bucket(given_date: date | datetime | str) -> tuple[int, int]:
    target_date = _coerce_to_date(given_date)
    first_day_of_week = _monday_of_week(target_date)
    last_date_of_week = first_day_of_week + timedelta(days=6)
    return last_date_of_week.year, last_date_of_week.month


# Calendar buckets anchor on the Monday that starts the week.
def get_calendar_month_bucket(given_date: date | datetime | str) -> tuple[int, int]:
    target_date = _coerce_to_date(given_date)
    first_day_of_week = _monday_of_week(target_date)
    return first_day_of_week.year, first_day_of_week.month


def get_broadcast_weeks_in_range(
    start_date: date | datetime | str,
    end_date: date | datetime | str,
) -> list[dict[str, date]]:
    start_value = _coerce_to_date(start_date)
    end_value = _coerce_to_date(end_date)
    if start_value > end_value:
        raise ValueError("startDate must be on or before endDate")

    start_week = _monday_of_week(start_value)
    end_week = _monday_of_week(end_value)

    weeks: list[dict[str, date]] = []
    cursor = start_week
    while cursor <= end_week:
        weeks.append(
            {
                "weekStart": cursor,
                "weekEnd": cursor + timedelta(days=6),
            }
        )
        cursor += timedelta(days=7)
    return weeks


def get_broadcast_calendar_info(given_date: date | datetime | str) -> dict[str, Any]:
    target_date = _coerce_to_date(given_date)

    first_day_of_week = _monday_of_week(target_date)
    last_date_of_week = first_day_of_week + timedelta(days=6)

    broadcast_year, broadcast_month = get_broadcast_month_bucket(target_date)
    calendar_year, calendar_month = get_calendar_month_bucket(target_date)

    begin_broadcast_month, end_broadcast_month = get_broadcast_month_range(
        broadcast_month,
        broadcast_year,
    )
    begin_calendar_month, end_calendar_month = get_calendar_month_range(
        calendar_month,
        calendar_year,
    )

    week_num_of_month = ((last_date_of_week - begin_broadcast_month).days + 1) // 7
    calendar_week_num_of_month = ((first_day_of_week - begin_calendar_month).days // 7) + 1

    broadcast_year_day_1 = date(broadcast_year, 1, 1)
    begin_broadcast_year = _monday_of_week(broadcast_year_day_1)
    week_num_of_year = ((last_date_of_week - begin_broadcast_year).days + 1) // 7

    calendar_year_day_1 = date(calendar_year, 1, 1)
    begin_calendar_year = _monday_of_week(calendar_year_day_1)
    calendar_week_num_of_year = ((first_day_of_week - begin_calendar_year).days // 7) + 1

    number_of_broadcast_week = ((end_broadcast_month - begin_broadcast_month).days + 1) // 7
    number_of_calendar_week = ((end_calendar_month - begin_calendar_month).days + 1) // 7

    return {
        "broadcastMonth": broadcast_month,
        "broadcastYear": broadcast_year,
        "beginBroadcastMonth": begin_broadcast_month,
        "endBroadcastMonth": end_broadcast_month,
        "numberOfBroadcastWeek": number_of_broadcast_week,
        "calendarMonth": calendar_month,
        "calendarYear": calendar_year,
        "beginCalendarMonth": begin_calendar_month,
        "endCalendarMonth": end_calendar_month,
        "numberOfCalendarWeek": number_of_calendar_week,
        "firstDayOfWeek": first_day_of_week,
        "lastDateOfWeek": last_date_of_week,
        "weekNumofMonth": week_num_of_month,
        "weekNumofYear": week_num_of_year,
        "calendarWeekNumofMonth": calendar_week_num_of_month,
        "calendarWeekNumofYear": calendar_week_num_of_year,
    }


def get_broadcast_calendar_value(
    given_date: date | datetime | str,
    result_type: str,
) -> Any:
    info = get_broadcast_calendar_info(given_date)
    normalized = str(result_type or "").strip().lower()
    result_type_map = {
        "month": "broadcastMonth",
        "year": "broadcastYear",
        "start_date": "beginBroadcastMonth",
        "end_date": "endBroadcastMonth",
        "num_of_week": "numberOfBroadcastWeek",
        "firstdate_of_week": "firstDayOfWeek",
        "lastdate_of_week": "lastDateOfWeek",
        "week_num_of_month": "weekNumofMonth",
        "week_num_of_year": "weekNumofYear",
        "calendar_month": "calendarMonth",
        "calendar_year": "calendarYear",
        "calendar_start_date": "beginCalendarMonth",
        "calendar_end_date": "endCalendarMonth",
        "calendar_num_of_week": "numberOfCalendarWeek",
        "calendar_firstdate_of_week": "firstDayOfWeek",
        "calendar_lastdate_of_week": "lastDateOfWeek",
        "calendar_week_num_of_month": "calendarWeekNumofMonth",
        "calendar_week_num_of_year": "calendarWeekNumofYear",
    }

    mapped_key = result_type_map.get(normalized)
    if not mapped_key:
        supported = ", ".join(result_type_map.keys())
        raise ValueError(
            f"Unsupported resultType '{result_type}'. Use one of: {supported}"
        )
    return info[mapped_key]

File: frontend/apps/tradsphere/src/lib/broadcastCalendar.ts
export const TRADSPHERE_BROADCAST_TIMEZONE = "America/Chicago";

export type BroadcastRange = {
  flightStart: string;
  flightEnd: string;
};

export type BroadcastCalendarInfo = {
  broadcastMonth: number;
  broadcastYear: number;
  beginBroadcastMonth: string;
  endBroadcastMonth: string;
  numberOfBroadcastWeek: number;
  calendarMonth: number;
  calendarYear: number;
  beginCalendarMonth: string;
  endCalendarMonth: string;
  numberOfCalendarWeek: number;
  firstDayOfWeek: string;
  lastDateOfWeek: string;
  weekNumofMonth: number;
  weekNumofYear: number;
  calendarWeekNumofMonth: number;
  calendarWeekNumofYear: number;
};

const MIN_SUPPORTED_YEAR = 1901;
const MAX_SUPPORTED_YEAR = 2155;

function createUtcDate(year: number, month: number, day: number): Date {
  return new Date(Date.UTC(year, month - 1, day));
}

function formatUtcDateToIso(dateValue: Date): string {
  const year = dateValue.getUTCFullYear();
  const month = String(dateValue.getUTCMonth() + 1).padStart(2, "0");
  const day = String(dateValue.getUTCDate()).padStart(2, "0");
  return `${year}-${month}-${day}`;
}

function addDays(dateValue: Date, days: number): Date {
  const next = new Date(dateValue.getTime());
  next.setUTCDate(next.getUTCDate() + days);
  return next;
}

function mondayOfWeek(dateValue: Date): Date {
  return addDays(dateValue, -dateValue.getUTCDay() + (dateValue.getUTCDay() === 0 ? -6 : 1));
}

function getLastDayOfMonth(year: number, month: number): number {
  return new Date(Date.UTC(year, month, 0)).getUTCDate();
}

function calendarMonthStart(year: number, month: number): Date {
  const monthStart = createUtcDate(year, month, 1);
  if (monthStart.getUTCDay() === 1) {
    return monthStart;
  }
  const offset = 7 - monthStart.getUTCDay();
  return addDays(monthStart, offset);
}

function calendarMonthEnd(year: number, month: number): Date {
  const monthEnd = createUtcDate(year, month, getLastDayOfMonth(year, month));
  return addDays(mondayOfWeek(monthEnd), 6);
}

function ensureYear(year: number): void {
  if (!Number.isInteger(year) || year < MIN_SUPPORTED_YEAR || year > MAX_SUPPORTED_YEAR) {
    throw new Error(`Year must be between ${MIN_SUPPORTED_YEAR} and ${MAX_SUPPORTED_YEAR}.`);
  }
}

function ensureMonth(month: number): void {
  if (!Number.isInteger(month) || month < 1 || month > 12) {
    throw new Error("Month must be between 1 and 12.");
  }
}

function ensureQuarter(quarter: number): void {
  if (!Number.isInteger(quarter) || quarter < 1 || quarter > 4) {
    throw new Error("Quarter must be between 1 and 4.");
  }
}

// Broadcast buckets anchor on the week-ending Sunday.
export function getBroadcastMonthRange(month: number, year: number): BroadcastRange {
  ensureMonth(month);
  ensureYear(year);

  const monthStart = createUtcDate(year, month, 1);
  const monthEnd = createUtcDate(year, month, getLastDayOfMonth(year, month));
  const beginBroadcastMonth = mondayOfWeek(monthStart);

  const diffDays = Math.floor((monthEnd.getTime() - beginBroadcastMonth.getTime()) / 86400000) + 1;
  const numberOfBroadcastWeek = Math.floor(diffDays / 7);

  if (numberOfBroadcastWeek < 1) {
    throw new Error("Unable to compute broadcast month range.");
  }

  const endBroadcastMonth = addDays(beginBroadcastMonth, numberOfBroadcastWeek * 7 - 1);
  return {
    flightStart: formatUtcDateToIso(beginBroadcastMonth),
    flightEnd: formatUtcDateToIso(endBroadcastMonth),
  };
}

// Calendar buckets full weeks by the Monday that starts the week.
export function getCalendarMonthRange(month: number, year: number): BroadcastRange {
  ensureMonth(month);
  ensureYear(year);

  const beginCalendarMonth = calendarMonthStart(year, month);
  const endCalendarMonth = calendarMonthEnd(year, month);

  return {
    flightStart: formatUtcDateToIso(beginCalendarMonth),
    flightEnd: formatUtcDateToIso(endCalendarMonth),
  };
}

export function getBroadcastQuarterRange(quarter: 1 | 2 | 3 | 4, year: number): BroadcastRange {
  ensureQuarter(quarter);
  ensureYear(year);

  const startMonth = (quarter - 1) * 3 + 1;
  const firstMonthRange = getBroadcastMonthRange(startMonth, year);
  const lastMonthRange = getBroadcastMonthRange(startMonth + 2, year);

  return {
    flightStart: firstMonthRange.flightStart,
    flightEnd: lastMonthRange.flightEnd,
  };
}

export function getCalendarQuarterRange(quarter: 1 | 2 | 3 | 4, year: number): BroadcastRange {
  ensureQuarter(quarter);
  ensureYear(year);

  const startMonth = (quarter - 1) * 3 + 1;
  const firstMonthRange = getCalendarMonthRange(startMonth, year);
  const lastMonthRange = getCalendarMonthRange(startMonth + 2, year);

  return {
    flightStart: firstMonthRange.flightStart,
    flightEnd: lastMonthRange.flightEnd,
  };
}

export function getBroadcastYearRange(year: number): BroadcastRange {
  ensureYear(year);

  const firstMonthRange = getBroadcastMonthRange(1, year);
  const lastMonthRange = getBroadcastMonthRange(12, year);

  return {
    flightStart: firstMonthRange.flightStart,
    flightEnd: lastMonthRange.flightEnd,
  };
}

export function getCalendarYearRange(year: number): BroadcastRange {
  ensureYear(year);

  const firstMonthRange = getCalendarMonthRange(1, year);
  const lastMonthRange = getCalendarMonthRange(12, year);

  return {
    flightStart: firstMonthRange.flightStart,
    flightEnd: lastMonthRange.flightEnd,
  };
}

export function getBroadcastMonthBucket(dateValue: Date): { year: number; month: number } {
  const sunday = addDays(mondayOfWeek(dateValue), 6);
  return {
    year: sunday.getUTCFullYear(),
    month: sunday.getUTCMonth() + 1,
  };
}

export function getCalendarMonthBucket(dateValue: Date): { year: number; month: number } {
  const monday = mondayOfWeek(dateValue);
  return {
    year: monday.getUTCFullYear(),
    month: monday.getUTCMonth() + 1,
  };
}

export function buildBroadcastYearOptions(centerYear: number, totalCount = 9): number[] {
  const safeCenter = Number.isInteger(centerYear) ? centerYear : new Date().getUTCFullYear();
  const half = Math.floor(totalCount / 2);
  const years: number[] = [];

  for (let offset = -half; offset <= half; offset += 1) {
    const year = safeCenter + offset;
    if (year < MIN_SUPPORTED_YEAR || year > MAX_SUPPORTED_YEAR) {
      continue;
    }
    years.push(year);
  }

  return years;
}
